x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -129,83 +161,103 @@ func (k Keeper) UpdateDelegationBalances(ctx sdk.Context, hostZone types.HostZon
-//  1. Update the time of each hostZoneUnbonding on each epochUnbondingRecord
-//  2. Return the number of stTokens that need to be burned
-func (k Keeper) UpdateHostZoneUnbondings(
+// Updates the host zone unbonding records after a successful undelegation batch
+// The StTokensToBurn and the NativeTokensToUnbond amounts on the records are
+// decremented in a cascading fashion starting from the earliest record
+// The latest completion times is also set on each record if the time from the
+// batch is later than what's currently on the record
+func (k Keeper) UpdateHostZoneUnbondingsAfterUndelegation(
 	ctx sdk.Context,
-	latestCompletionTime time.Time,
 	chainId string,
-	undelegateCallback types.UndelegateCallback,
-) (stTokenBurnAmount sdkmath.Int, err error) {
-	stTokenBurnAmount = sdkmath.ZeroInt()
-	for _, epochNumber := range undelegateCallback.EpochUnbondingRecordIds {
-		epochUnbondingRecord, found := k.RecordsKeeper.GetEpochUnbondingRecord(ctx, epochNumber)
+	epochUnbondingRecordIds []uint64,
+	totalStTokensToBurn sdkmath.Int,
+	totalNativeTokensUnbonded sdkmath.Int,
+	unbondingTime uint64,
+) error {
+	// Loop each epoch unbonding record starting from the earliest
+	for _, epochNumber := range epochUnbondingRecordIds {
+		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochNumber, chainId)
 		if !found {
-			errMsg := fmt.Sprintf("Unable to find epoch unbonding record for epoch: %d", epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+			return errorsmod.Wrapf(recordstypes.ErrHostUnbondingRecordNotFound,
+				"host zone unbonding not found for epoch %d and %s", epochNumber, chainId)
 		}
-		hostZoneUnbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, epochUnbondingRecord.EpochNumber, chainId)
-		if !found {
-			errMsg := fmt.Sprintf("Host zone unbonding not found (%s) in epoch unbonding record: %d", chainId, epochNumber)
-			k.Logger(ctx).Error(errMsg)
-			return sdkmath.ZeroInt(), errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+
+		// Determine the amount to decrement from each record, flooring at 0
+		stTokensToBurn := sdkmath.MinInt(hostZoneUnbonding.StTokensToBurn, totalStTokensToBurn)
+		nativeTokensUnbonded := sdkmath.MinInt(hostZoneUnbonding.NativeTokensToUnbond, totalNativeTokensUnbonded)
+
+		// Decrement the records
+		hostZoneUnbonding.StTokensToBurn = hostZoneUnbonding.StTokensToBurn.Sub(stTokensToBurn)
+		hostZoneUnbonding.NativeTokensToUnbond = hostZoneUnbonding.NativeTokensToUnbond.Sub(nativeTokensUnbonded)
+
+		// Update the totals for the next loop
+		totalStTokensToBurn = totalStTokensToBurn.Sub(stTokensToBurn)
+		totalNativeTokensUnbonded = totalNativeTokensUnbonded.Sub(nativeTokensUnbonded)
+
+		// If there are no more tokens to burn after this batch, iterate the record to the next status
+		if hostZoneUnbonding.StTokensToBurn.IsZero() && hostZoneUnbonding.NativeTokensToUnbond.IsZero() {
```

Comment:
Is this where you would need to take into account the off by one rounding errors instead of absolute comparison to zero?

Reply:
this is actually from your point in the other PR! since we have bended redemption rates across records, there's no guarantee that they will zero out at the same time, so we have to wait for both to be zero before we iterate the record. 

[Here's](https://github.com/Stride-Labs/stride/pull/1198/files#diff-b534601426c987137850288cf1ff0c31eec9ba0127489956e0399974b15dcf33R582-R584) a unit test that cover's this situation

Reply:
I could definitely be misunderstanding, but what's the case where `NativeTokensToUnbond > 0` and we don't want to iterate the record?

Reply:
After completing the first batch, the `NativeTokensToUnbond` will likely be still greater than 0, so we want to keep the record in `IN_PROGRESS` until enough batches complete that it decrements all the tokens on that record